### PR TITLE
First stab at making environmentServiceName optional

### PR DESCRIPTION
**Pull request by favoretti**
Addresses #54

@charleszipp - please review. Still pondering what to do with tests.

### CODE REVIEW

File: TerraformCLI/src/terraform-runner.ts [hunk 1]
```diff
@@ -37,18 +37,18 @@ export class TerraformWithCommand extends TerraformCommandBuilder{
...
-    private readonly environmentServiceName: string;
-    constructor(builder: TerraformCommandBuilder, environmentServiceName: string) {
+export class TerraformWithProvider extends TerraformCommandDecorator{
+    private readonly environmentServiceName: string | undefined;
+    constructor(builder: TerraformCommandBuilder, environmentServiceName: string | undefined) {
         super(builder);
         this.environmentServiceName = environmentServiceName;
     }
-    async onRun(context: TerraformCommandContext): Promise<void> {        
+    async onRun(context: TerraformCommandContext): Promise<void> {
```
Comment (charleszipp) [inline TerraformCLI/src/terraform-runner.ts:51]: For the question regarding tests, lets add a test that doesnt setup the `environmentServiceName` input. The test should at least verify the task succeeds when this value is not provided. See example below. Lets add one test scenario that does this (can be copied from vanilla azurerm scenario) to apply, plan, destroy, import, and refresh.

```javascript
new TaskScenario<TerraformInputs>()
    .inputAzureRmServiceEndpoint(env.environmentServiceName, env.subscriptionId, env.tenantId, env.servicePrincipalId, env.servicePrincipalKey)
    .inputTerraformCommand(env.terraformCommand)
    // do not provide environmentServiceName input
    /// .input({ environmentServiceName: env.environmentServiceName })
    .inputApplicationInsightsInstrumentationKey()
    .answerTerraformExists()
    .answerTerraformCommandIsSuccessful(env.commandArgs)
    .answerTerraformCommandWithVarsFileAsWorkingDirFails()
    .run()
```

File: TerraformCLI/task.json [hunk 2]
```diff
@@ -76,10 +76,10 @@
         {
             "name": "environmentServiceName",
             "type": "connectedService:AzureRM",
-            "label": "Environment Azure Subscription",
+            "label": "Environment Azure Subscription (select if not configured inside terraform code)",
             "required": false,
             "visibleRule": "command = plan || command = apply || command = destroy || command = refresh || command = import",
-            "helpMarkDown": "Select an Azure resource manager subscription for the deployment"
+            "helpMarkDown": "Select an Azure resource manager subscription for the deployment. If you configure provider using secureEnvFile, you can leave this empty."
```
Comment (charleszipp) [inline TerraformCLI/task.json:82]: Update this to the following. Be sure to include the markdown syntax.

> Select an Azure service connection if you would like the `azurerm` provider configured for you (using the service principal credentials configured in the selected service connection). **This field is optional.** Leave empty should you choose to self-configure the provider(s) used. Provider environment variables can be configured securely by selecting an env file from the `Secured Variables File` pick list.

File: TerraformCLI/src/terraform-runner.ts (hunk 1 above)
Comment (favoretti) [reply TerraformCLI/src/terraform-runner.ts:51]: @charleszipp I guess we don't need `.inputAzureRmServiceEndpoint(env.environmentServiceName, env.subscriptionId, env.tenantId, env.servicePrincipalId, env.servicePrincipalKey)` either then?

**Comment by favoretti** [comment]
@charleszipp added tests (I hope my assumption on not needing the service connection was correct, let me know if not, I'll put it back :)), fixed description.

File: TerraformCLI/src/terraform-runner.ts (hunk 1 above)
Comment (charleszipp) [reply TerraformCLI/src/terraform-runner.ts:51]: Correct. One is simulating the setup/existence of the service connection (i.e. via the project settings) the basic input one is simulating the user selecting the service connection for use by the task.

File: TerraformCLI/src/tests/terraform-apply/apply_spec.ts [hunk 3]
```diff
@@ -2,61 +2,71 @@ import { TestScenario } from '../assertions';
...
         let env = require('./apply-azurerm-with-invalid-auth-scheme.env').env;
         new TestScenario(env.taskScenarioPath)
             .assertExecutionFailed('Terraform only supports service principal authorization for azure')
             .assertExecutedTerraformVersion()
             // test runner does not expose env vars set within the task so cannot use this yet
             //.andAssert((assertions) => new TaskExecutedWithEnvironmentVariables(assertions, env.expectedEnv));
             .run();
+    });
+    it('without envservicename', function(){
+        let env = require('./apply-without-envservicename.env').env;
```
Comment (charleszipp) [inline TerraformCLI/src/tests/terraform-apply/apply_spec.ts:63]: Did these get left out of your commit? I am not seeing them in the changes

**Comment by charleszipp** [comment]
@favoretti the build is failing because the tests are failing. Run `npm run test:local` to run the tests to see which are failing. There is also a launch configuration for debugging the tests (one for individual scenario too).

File: TerraformCLI/src/tests/terraform-apply/apply_spec.ts (hunk 3 above)
Comment (favoretti) [reply TerraformCLI/src/tests/terraform-apply/apply_spec.ts:63]: Joys of multitasking... Sorry about that.

**Comment by favoretti** [comment]
> @favoretti the build is failing because the tests are failing. Run `npm run test:local` to run the tests to see which are failing. There is also a launch configuration for debugging the tests (one for individual scenario too).

Yeah, I'm an idiot, didn't add the new files. Fixed now :)

**Pull request merged**